File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleops/arcade_drive/ArcadeDriveBase.java
package org.firstinspires.ftc.teamcode.teleops.arcade_drive;

import com.qualcomm.robotcore.util.Range;

import org.firstinspires.ftc.teamcode.teleops.TeleopBase;

abstract public class ArcadeDriveBase extends TeleopBase {
    @Override
    final protected void opModeCode() throws InterruptedException {
        hw.setGrabberIsOpen(false);

        waitForStart();

        while (opModeIsActive()) {
            if (aGamepad2.xPressed()) hw.resetGrabberEncoder();

            // drive
            float left = Range.clip(aGamepad1.leftY() + aGamepad1.leftX(), -1, 1);
            float right = Range.clip(aGamepad1.leftY() - aGamepad1.leftX(), -1, 1);
            hw.setAdjustedWheelPower(left, right);

            // duck spinner
            hw.setSpinnerPower(Math.max(-gamepad2.left_stick_y, 0), getSide());

            // grabber
            if (aGamepad2.rightBumperPressed()) hw.toggleGrabberState();
            if (aGamepad2.dpadUpPressed()) hw.raiseGrabberHeight();
            if (aGamepad2.dpadDownPressed()) hw.lowerGrabberHeight();
            hw.setGrabberOffset((int) (gamepad2.left_trigger * 50));
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleops/arcade_drive/ArcadeDriveBase.kt
package org.firstinspires.ftc.teamcode.teleops.arcade_drive

import org.firstinspires.ftc.teamcode.utils.Side
import org.firstinspires.ftc.teamcode.teleops.TeleopBase
import kotlin.math.abs

abstract class ArcadeDriveBase : TeleopBase() {
    abstract val side: Side

    override fun beforeStart() {
        hw.grabberIsOpen = false
    }

    override fun mainOpMode() {
        while (opModeIsActive()) {
            if (aGamepad2.xPressed) hw.resetGrabberEncoder()

            var leftPower = aGamepad1.leftY + aGamepad1.rightX
            var rightPower = aGamepad1.leftY - aGamepad1.rightX
            if ((leftPower > rightPower && leftPower > 1) || (leftPower < rightPower && leftPower < -1)) {
                val factor = abs(leftPower)
                leftPower /= factor
                rightPower /= factor
            } else if ((rightPower > leftPower && rightPower > 1) || (rightPower < leftPower && rightPower < -1)) {
                val factor = abs(rightPower)
                leftPower /= factor
                rightPower /= factor
            }

            // drive
            hw.setAdjustedWheelPower(leftPower, rightPower)

            // duck spinner
            hw.setSpinnerPower((-aGamepad2.leftY).coerceAtLeast(0.0), side)

            // grabber
            if (aGamepad2.rightBumperPressed) hw.toggleGrabberState()
            if (aGamepad2.dpadUpPressed) hw.raiseGrabberHeight()
            if (aGamepad2.dpadDownPressed) hw.lowerGrabberHeight()
            hw.grabberOffset = (gamepad2.left_trigger * 50).toInt()
        }
    }
}